sklearn/cluster/_birch.py

```diff
@@ -284,7 +284,8 @@ def __init__(self, *, linear_sum=None):
             self.centroid_ = self.linear_sum_ = 0
         else:
             self.n_samples_ = 1
-            self.centroid_ = self.linear_sum_ = linear_sum
+            # defensive copy and ensure FP64 precision because of numerical issues
+            self.centroid_ = self.linear_sum_ = np.array(linear_sum, dtype=np.float64)
```

Comment:
Reading this line, I think having `self.centroid_` as a reference to the same array as `self.linear_sum_` could lead to bugs. I think this does not happen because `update(self, subcluster)` method below that fixes `self.centroid_` by reallocating it. Still this is confusing. I think it should be:

```suggestion
            self.linear_sum_ = linear_sum.astype(np.float64, copy=False)
            self.centroid_ = self.linear_sum_.copy()  # centroid of 1 sample
```

Reply:
Or maybe this is an intentional optimization... In which case we could just add a comment.

Reply:
Given the code quality of birch here, I doubt this is an intentional optimization.
I would use `copy=True`, because what if the input array X gets modified and some leaf only has a single element?
It will likely be more efficient to reuse the arrays in `update` then than to allocate new one each time.